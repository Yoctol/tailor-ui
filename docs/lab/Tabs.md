---
id: tabs
title: Tabs
---

Tabs make it easy to switch between different views.

## When To Use

Use tabs for functional aspects of a page.

## Examples

```js
import { Tabs } from 'tailor-ui';
```

### Basic

```jsx live
() => {
  const [value, setValue] = useState('1');

  return (
    <Box bg="primaryDark2" p="3" borderRadius="xl">
      <Tabs value={value} onChange={setValue}>
        <Tabs.Tab value="1">Tab 1</Tabs.Tab>
        <Tabs.Tab value="2">Tab 2</Tabs.Tab>
        <Tabs.Tab value="3">Tab 3</Tabs.Tab>
      </Tabs>

      <Box color="light">
        <p>Active Tab is {value}</p>
      </Box>
    </Box>
  );
}
```

### With size

```jsx live
<Box bg="primaryDark2" p="3" borderRadius="xl">
  <Tabs size="sm" defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>

  <br />

  <Tabs defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>

  <br />

  <Tabs size="lg" defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>
</Box>
```

### With controlled activeValue

```jsx live
() => {
  const [value, setValue] = useState('1');

  return (
    <Box bg="primaryDark2" p="3" borderRadius="xl">
      <Tabs value={value} onChange={setValue}>
        <Tabs.Tab value="1">Tab 1</Tabs.Tab>
        <Tabs.Tab value="2">Tab 2</Tabs.Tab>
        <Tabs.Tab value="3">Tab 3</Tabs.Tab>
      </Tabs>

      <br />

      <Tabs value={value} onChange={setValue}>
        <Tabs.Tab value="1">Tab 1</Tabs.Tab>
        <Tabs.Tab value="2">Tab 2</Tabs.Tab>
        <Tabs.Tab value="3">Tab 3</Tabs.Tab>
      </Tabs>
    </Box>
  );
}
```

### Card type tab

```jsx live
() => {
  const [value, setValue] = useState('1');

  return (
    <>
      <Tabs type="card" value={value} onChange={setValue}>
        <Tabs.Tab value="1">Tab 1</Tabs.Tab>
        <Tabs.Tab value="2">Tab 2</Tabs.Tab>
        <Tabs.Tab value="3">Tab 3</Tabs.Tab>
      </Tabs>

      <p>Active Tab is {value}</p>
    </>
  );
}
```

### Card type tab with size

```jsx live
<>
  <Tabs type="card" size="sm" defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>

  <br />

  <Tabs type="card" defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>

  <br />

  <Tabs type="card" size="lg" defaultValue="1">
    <Tabs.Tab value="1">Tab 1</Tabs.Tab>
    <Tabs.Tab value="2">Tab 2</Tabs.Tab>
    <Tabs.Tab value="3">Tab 3</Tabs.Tab>
  </Tabs>
</>
```

## API

### Tabs

| Property       | Description                                              | Type                            | Default  |
|----------------|----------------------------------------------------------|---------------------------------|----------|
| `defaultValue` | Initial active Tab's value, if `activeValue` is not set. | `string`                        |          |
| `value`        | Current Tab's value                                      | `string`                        |          |
| `onChange`     | Callback executed when active tab is changed             | `(activeValue: string) => void` |          |
| `size`         | Preset tab bar size                                      | `'sm'` \| `'md'` \| `'lg'`      | `'md'`   |
| `type`         | Type of tab                                              | `'line'` \| `'card'`            | `'line'` |

### Tabs.Tab

| Property | Description | Type     | Default |
|----------|-------------|----------|---------|
| `value`  | Tab's value | `string` |         |
