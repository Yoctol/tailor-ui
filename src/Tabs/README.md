# Tabs Docs

```js
import { Tabs } from '@yoctol/ui';
```

### Try it with `knobs`:

<!-- STORY -->

## API

### Tabs

| Props      | type                                                                              | Default | Description                                    |
| ---------- | --------------------------------------------------------------------------------- | ------- | ---------------------------------------------- |
| `absolute` | Boolean                                                                           | `false` | Make the tabs position absolute on the bottom. |
| `block`    | Boolean                                                                           | `false` | Set the Tabs width to 100%.                    |
| `size`     | `'sm'`&#124;`'m'`&#124;`'lg'`                                                     | `'m'`   |                                                |
| Space      | [View on docs](https://github.com/jxnblk/styled-system/blob/master/docs/table.md) |         | From `styled-system`                           |

### Tabs.Tab

| Props      | type    | Default | Description |
| ---------- | ------- | ------- | ----------- |
| `active`   | Boolean | `false` |             |
| `disabled` | Boolean | `false` |             |
