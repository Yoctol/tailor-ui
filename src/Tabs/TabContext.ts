import { createContext } from 'react';

export type Size = 'sm' | 'md' | 'lg';

const TabContext = createContext<{
  activeValue: string;
  setValue: (value: string) => void;
  size: Size;
  pills: boolean;
}>({
  activeValue: '',
  setValue: () => {},
  size: 'md',
  pills: false,
});

export default TabContext;
