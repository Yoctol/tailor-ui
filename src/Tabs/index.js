import React, { Children, cloneElement } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { themeGet, space, borderRadius } from 'styled-system';
import { ifProp, switchProp } from 'styled-tools';

import theme from '../theme';

const PillsTab = css`
  border-radius: ${themeGet('radii.1')};
  background-color: ${ifProp(
    'active',
    themeGet('colors.gray.8'),
    themeGet('colors.bgLight')
  )};
`;

const DefaultTab = css`
  border-bottom: 3px solid transparent;

  ${ifProp(
    'active',
    css`
      border-color: ${themeGet('colors.secondary')};
    `
  )};
`;

const Tab = styled.a`
  position: relative;
  display: inline-block;
  text-align: center;
  cursor: pointer;
  background-color: ${ifProp('active', themeGet('colors.bgLight'))};

  &:hover {
    text-decoration: none;
  }

  &:disabled,
  &[disabled] {
    cursor: default;
    opacity: 0.5;
    pointer-events: none;
  }

  ${switchProp('size', {
    sm: css`
      height: ${themeGet('space.sizeSm')};
      padding: ${themeGet('space.paddingYSm')} ${themeGet('space.paddingXSm')};
      font-size: ${themeGet('fontSizes.sm')};
    `,

    m: css`
      height: ${themeGet('space.size')};
      padding: ${themeGet('space.paddingY')} ${themeGet('space.paddingX')};
      font-size: ${themeGet('fontSizes.default')};
    `,

    lg: css`
      height: ${themeGet('space.sizeLg')};
      padding: ${themeGet('space.paddingYLg')} ${themeGet('space.paddingXLg')};
      font-size: ${themeGet('fontSizes.lg')};
    `,
  })};

  ${ifProp('pills', PillsTab, DefaultTab)};

  ${space};
  ${borderRadius};
  text-decoration: none;
`;

Tab.propTypes = {
  active: PropTypes.bool,
};

Tab.defaultProps = {
  theme,
  active: false,
  size: 'm',
};

const TabsWrapper = styled.nav`
  display: flex;

  ${ifProp(
    { pills: false },
    css`
      position: absolute;
      bottom: -3px;
    `
  )};

  /* stylelint-disable no-duplicate-selectors */
  ${Tab} {
    flex-grow: ${ifProp('block', 1, 0)};
  }
  /* stylelint-enable */

  ${space};
`;

const Tabs = ({ children, pills, size, ...otherProps }) => (
  <TabsWrapper pills={pills} {...otherProps}>
    {Children.map(children, child => cloneElement(child, { pills, size }))}
  </TabsWrapper>
);

Tabs.propTypes = {
  block: PropTypes.bool,
  children: PropTypes.node.isRequired,
  pills: PropTypes.bool,
  ...space.propTypes,
};

Tabs.defaultProps = {
  theme, // eslint-disable-line react/default-props-match-prop-types
  block: false,
  pills: false,
};

Tabs.Tab = Tab;

export default Tabs;
