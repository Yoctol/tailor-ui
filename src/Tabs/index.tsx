import React, { PureComponent, SFC, createContext } from 'react';
import { Value } from 'react-powerplug';
import { space as styledSpace } from 'styled-system';

import styled, { css } from 'utils/styled-components';

type Size = 'sm' | 'md' | 'lg';

const { Provider, Consumer } = createContext<{
  activeValue: string;
  setValue: (value: string) => void;
  size: Size;
  pills: boolean;
}>({
  activeValue: '',
  setValue: () => {},
  size: 'md',
  pills: false,
});

interface StyledTabProps {
  size: Size;
  pills: boolean;
  active: boolean;
}

const StyledTab = styled<StyledTabProps, 'a'>('a')`
  display: inline-flex;
  position: relative;
  align-items: center;
  color: ${p => p.theme.colors.dark};
  text-align: center;
  text-decoration: none;
  cursor: pointer;

  &:hover {
    text-decoration: none;
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  &:not(:first-child) {
    margin-left: ${p => p.theme.space[2]};
  }

  ${({ size, theme: { heights, paddings, fontSizes } }) =>
    ({
      sm: css`
        height: ${heights.sm};
        padding: 0 ${paddings.sm};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 0 ${paddings.md};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 0 ${paddings.lg};
        font-size: ${fontSizes.lg};
      `,
    }[size])};

  ${({ pills, active }) =>
    pills
      ? css`
          border-radius: 999px;
          ${active &&
            css`
              background-color: ${p => p.theme.colors.primary};
              color: ${p => p.theme.colors.light};
            `};
        `
      : css`
          border-bottom: 3px solid transparent;

          ${active &&
            css`
              border-bottom-color: ${p => p.theme.colors.secondary};
            `};
        `};

  ${p => p.theme.transition /* sc-declaration */};
  ${styledSpace};
`;

export type TabProps = {
  label: string;
  value: string;
  disabled?: boolean;
};

export const Tab: SFC<TabProps> = ({ label, value, ...props }) => (
  <Consumer>
    {({ size, pills, activeValue, setValue }) => (
      <StyledTab
        size={size}
        pills={pills}
        active={activeValue === value}
        onClick={() => setValue(value)}
        {...props}
      >
        {label}
      </StyledTab>
    )}
  </Consumer>
);

// Tab.propTypes = {
//   /**
//    * Disabled state of tab
//    */
//   disabled: PropTypes.bool,
//   /**
//    * Content of tab
//    */
//   label: PropTypes.node,
//   /**
//    * Tab's value
//    */
//   value: PropTypes.string,
// };

// Tab.defaultProps = {
//   label: '',
//   disabled: false,
//   value: '',
// };

interface StyledTabsProps {
  /**
   * Make the tabs position to bottom of parent
   */
  absolute?: boolean;
  /**
   * Set the tabs width to 100%
   */
  block?: boolean;
}

const StyledTabs = styled<StyledTabsProps, 'nav'>('nav')`
  display: flex;

  ${({ absolute }) =>
    absolute &&
    css`
      position: absolute;
      bottom: -3px;
    `};

  ${({ block }) =>
    block &&
    css`
      right: 0;
      left: 0;
      padding-right: inherit;
      padding-left: inherit;
      ${StyledTab /* sc-selector */} {
        flex-grow: 1;
      }
    `};

  ${styledSpace};
`;

type TabsProps = StyledTabsProps & {
  /**
   * Initial active Tab's value, if `activeValue` is not set.
   */
  defaultActiveValue?: string;
  /**
   * Current Tab's value
   */
  activeValue?: string;
  /**
   * Callback executed when active tab is changed
   */
  onChange?: (activeValue: string) => void;
  /**
   * Pills style of tabs
   */
  pills?: boolean;
  /**
   * Preset tab bar size
   */
  size?: Size;
};

class Tabs extends PureComponent<TabsProps> {
  static Tab = Tab;

  render() {
    const {
      absolute = false,
      pills = false,
      size = 'md',
      children,
      defaultActiveValue,
      activeValue,
      onChange,
      ...otherProps
    } = this.props;

    return (
      <Value
        initial={defaultActiveValue}
        onChange={value => {
          if (onChange && value) {
            onChange(value);
          }
        }}
      >
        {({ value, set }) => (
          <StyledTabs absolute={absolute} {...otherProps}>
            <Provider
              value={{
                activeValue: activeValue || value || '',
                setValue: set,
                size,
                pills,
              }}
            >
              {children}
            </Provider>
          </StyledTabs>
        )}
      </Value>
    );
  }
}

export default Tabs;
