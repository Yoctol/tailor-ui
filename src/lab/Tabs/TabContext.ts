import { MutableRefObject, createContext } from 'react';

const TabContext = createContext<{
  activeTabRef: MutableRefObject<HTMLAnchorElement | null>;
  activeValue: string;
  setValue: (value: string) => void;
}>({
  activeTabRef: { current: null },
  activeValue: '',
  setValue: () => {},
});

TabContext.displayName = 'TabContext';

export default TabContext;
