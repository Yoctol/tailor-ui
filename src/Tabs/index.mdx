---
name: Tabs
menu: Components
route: /components/tabs
---

import { PropsTable, Playground } from 'docz';
import { Space, Box } from '../';
import Tabs, { Tab } from './';

# Tabs

Tabs make it easy to switch between different views.

## When To Use

Use tabs for functional aspects of a page.

## Examples

```js
import { Tabs } from 'tailor-ui';
```

### Basic

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <>
        <Tabs activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <p>Active Tab is {value}</p>
      </>
    );

}}

</Playground>

### With pills

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <>
        <Tabs pills activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <p>Active Tab is {value}</p>
      </>
    );

}}

</Playground>

### With absolute

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <Box
        position="relative"
        height="120px"
        width="300px"
        borderBottom="3px solid #ddd"
      >
        <Tabs absolute activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <p>Active Tab is {value}</p>
      </Box>
    );

}}

</Playground>

### With disabled

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <>
        <Tabs activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" disabled label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <p>Active Tab is {value}</p>
      </>
    );

}}

</Playground>

### With pills & disabled

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <>
        <Tabs pills activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" disabled label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <p>Active Tab is {value}</p>
      </>
    );

}}

</Playground>

### With size

<Playground>
  <Tabs size="sm" defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
  <Tabs defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
  <Tabs size="lg" defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
</Playground>

### With size & pills

<Playground>
  <Tabs pills size="sm" defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
  <Tabs pills defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
  <Tabs pills size="lg" defaultActiveValue="1">
    <Tabs.Tab value="1" label="Tab 1" />
    <Tabs.Tab value="2" label="Tab 2" />
    <Tabs.Tab value="3" label="Tab 3" />
  </Tabs>
</Playground>

### With controlled activeValue

<Playground>
  {() => {
    const [value, setValue] = React.useState('1');

    return (
      <>
        <Tabs pills activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
        <Space my={2} />
        <Tabs pills activeValue={value} onChange={setValue}>
          <Tabs.Tab value="1" label="Tab 1" />
          <Tabs.Tab value="2" label="Tab 2" />
          <Tabs.Tab value="3" label="Tab 3" />
        </Tabs>
      </>
    );

}}

</Playground>

## API

### Tabs

<PropsTable of={Tabs} />

### Tabs.Tab

<PropsTable of={Tab} />
