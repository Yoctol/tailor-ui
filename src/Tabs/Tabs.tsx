import React, { FC, useContext, useState } from 'react';
import styled, { css } from 'styled-components';
import { SpaceProps, space } from 'styled-system';

import TabContext, { Size } from './TabContext';

interface StyledTabProps extends SpaceProps {
  size: Size;
  pills: boolean;
  active: boolean;
}

const StyledTab = styled.a<StyledTabProps>`
  display: inline-flex;
  position: relative;
  align-items: center;
  color: ${(p) => p.theme.colors.gray800};
  text-align: center;
  text-decoration: none;
  cursor: pointer;

  &:hover {
    text-decoration: none;
  }

  &:disabled,
  &[disabled] {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
  }

  &:not(:first-child) {
    margin-left: ${(p) => p.theme.space[2]};
  }

  ${({
    size,
    theme: { heights, paddings, fontSizes },
  }: StyledTabProps & { theme: any }) =>
    ({
      sm: css`
        height: ${heights.sm};
        padding: 0 ${paddings.sm};
        font-size: ${fontSizes.sm};
      `,
      md: css`
        height: ${heights.base};
        padding: 0 ${paddings.md};
        font-size: ${fontSizes.base};
      `,
      lg: css`
        height: ${heights.lg};
        padding: 0 ${paddings.lg};
        font-size: ${fontSizes.lg};
      `,
    }[size])};

  ${({ pills, active, theme }) =>
    pills
      ? css`
          border-radius: ${theme.radii.lg};
          color: ${theme.colors.gray600};

          ${active
            ? css`
                background-color: ${theme.colors.primary};
                color: ${theme.colors.light};
              `
            : css`
                &:hover {
                  background-color: ${theme.colors.gray300};
                  color: ${theme.colors.gray700};
                }
              `};
        `
      : css`
          border-bottom: 3px solid transparent;

          ${active &&
          css`
            border-bottom-color: ${theme.colors.primary};
          `};
        `};

  ${(p) => p.theme.transition /* sc-declaration */};
  ${space};
`;

export interface TabProps {
  /**
   * label of tab
   */
  label: string;
  /**
   * Tab's value
   */
  value: string;
  /**
   * Disabled state of tab
   */
  disabled?: boolean;
}

export const Tab: FC<TabProps> = ({ label, value, ...props }) => {
  const { size, pills, activeValue, setValue } = useContext(TabContext);

  return (
    <StyledTab
      size={size}
      pills={pills}
      active={activeValue === value}
      onClick={() => setValue(value)}
      {...props}
    >
      {label}
    </StyledTab>
  );
};

interface StyledTabsProps extends SpaceProps {
  /**
   * Make the tabs position to bottom of parent
   */
  absolute?: boolean;
  /**
   * Set the tabs width to 100%
   */
  block?: boolean;
}

const StyledTabs = styled.nav<StyledTabsProps>`
  display: flex;

  ${({ absolute }) =>
    absolute &&
    css`
      position: absolute;
      bottom: -3px;
    `};

  ${({ block }) =>
    block &&
    css`
      right: 0;
      left: 0;
      padding-right: inherit;
      padding-left: inherit;
      ${StyledTab /* sc-selector */} {
        flex-grow: 1;
      }
    `};

  ${space};
`;

type TabsProps = StyledTabsProps & {
  /**
   * Initial active Tab's value, if `activeValue` is not set.
   */
  defaultActiveValue?: string;
  /**
   * Current Tab's value
   */
  activeValue?: string;
  /**
   * Callback executed when active tab is changed
   */
  onChange?: (activeValue: string) => void;
  /**
   * Pills style of tabs
   */
  pills?: boolean;
  /**
   * Preset tab bar size
   */
  size?: Size;
};

const Tabs: FC<TabsProps> & {
  Tab: typeof Tab;
} = ({
  absolute = false,
  pills = false,
  size = 'md' as Size,
  children,
  defaultActiveValue,
  activeValue,
  onChange,
  ...otherProps
}) => {
  const [activeTab, setActiveTab] = useState(defaultActiveValue);

  return (
    <TabContext.Provider
      value={{
        activeValue: activeValue || activeTab || '',
        setValue: (value) => {
          setActiveTab(value);

          if (onChange && value) {
            onChange(value);
          }
        },
        size,
        pills,
      }}
    >
      <StyledTabs absolute={absolute} {...otherProps}>
        {children}
      </StyledTabs>
    </TabContext.Provider>
  );
};

Tabs.Tab = Tab;

export { Tabs };
