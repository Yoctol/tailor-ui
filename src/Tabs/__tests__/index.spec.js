import React from 'react';

import shallowWithTheme from 'test/shallowWithTheme';

import Tabs from '../';

const { Tab } = Tabs;

describe('Tabs', () => {
  it('should render correctly', () => {
    const component = shallowWithTheme(
      <Tabs>
        <Tab active>Tab 1</Tab>
        <Tab disabled>Tab 2</Tab>
        <Tab>Tab 3</Tab>
      </Tabs>
    );

    expect(component).toMatchSnapshot();
  });

  it('should render with props absolute', () => {
    const component = shallowWithTheme(
      <Tabs absolute>
        <Tab active>Tab 1</Tab>
        <Tab disabled>Tab 2</Tab>
        <Tab>Tab 3</Tab>
      </Tabs>
    );

    expect(component).toMatchSnapshot();
  });

  it('should render with props pills', () => {
    const component = shallowWithTheme(
      <Tabs pills>
        <Tab active>Tab 1</Tab>
        <Tab disabled>Tab 2</Tab>
        <Tab>Tab 3</Tab>
      </Tabs>
    );

    expect(component).toMatchSnapshot();
  });
});
