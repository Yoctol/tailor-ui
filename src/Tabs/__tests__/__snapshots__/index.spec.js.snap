// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render correctly 1`] = `
<Tabs__StyledTabs
  absolute={false}
  block={false}
>
  <[object Object]
    value="m"
  >
    <Tabs.Tab
      active={true}
      disabled={false}
      size="m"
    >
      Tab 1
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={true}
      size="m"
    >
      Tab 2
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={false}
      size="m"
    >
      Tab 3
    </Tabs.Tab>
  </[object Object]>
</Tabs__StyledTabs>
`;

exports[`Tabs should render with props absolute 1`] = `
<Tabs__StyledTabs
  absolute={true}
  block={false}
>
  <[object Object]
    value="m"
  >
    <Tabs.Tab
      active={true}
      disabled={false}
      size="m"
    >
      Tab 1
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={true}
      size="m"
    >
      Tab 2
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={false}
      size="m"
    >
      Tab 3
    </Tabs.Tab>
  </[object Object]>
</Tabs__StyledTabs>
`;

exports[`Tabs should render with props pills 1`] = `
<Tabs__StyledTabs
  absolute={false}
  block={false}
  pills={true}
>
  <[object Object]
    value="m"
  >
    <Tabs.Tab
      active={true}
      disabled={false}
      size="m"
    >
      Tab 1
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={true}
      size="m"
    >
      Tab 2
    </Tabs.Tab>
    <Tabs.Tab
      active={false}
      disabled={false}
      size="m"
    >
      Tab 3
    </Tabs.Tab>
  </[object Object]>
</Tabs__StyledTabs>
`;
